### PR TITLE
feat(typescript): add React Query subexport with options factories + hooks

### PR DESCRIPTION
**Pull request by Swimburger**
## Description
Linear ticket: Refs 

Refs customer request for React Query support in generated TypeScript SDKs.

Adds an opt-in `generateReactQuery: true` config option that generates TanStack Query v5 compatible helpers under the `./react-query` subexport. Generated code is organized into **per-service files** with barrel re-exports, enabling tree-shakeable subpath imports (e.g. `import { getUserOptions, useGetUser } from "@company/sdk/react-query/users"`).

**React-idiomatic DI**: Generated code includes a `ClientProvider` component and `useClient()` hook for React Context-based client injection. All generated hooks resolve the client from context automatically — no need to pass `client` as a parameter.

Link to Devin session: https://app.devin.ai/sessions/9089a03c15444984a9c02a9472f5953d  
Requested by: @Swimburger

### Generator
- Added `generateReactQuery` boolean config option (default `false`) to `TypescriptCustomConfigSchema`, `SdkCustomConfig`, and `SdkGeneratorCli`
- Created **`ReactQueryGenerator` class** (`react-query/ReactQueryGenerator.ts`) extracted from `SdkGenerator`, following existing generator class patterns. It handles:
  - Collecting all endpoint entries across packages
  - Grouping endpoints by service path
  - Generating per-service files with short function names (e.g. `getUserOptions` not `endpointsUsersGetUserOptions`)
  - Generating barrel `index.ts` files at each directory level with namespace re-exports
  - Generating `context.ts` with `ClientContext`, `ClientProvider` component, and `useClient()` hook
  - **Generating unit test files** (`tests/unit/react-query/*.test.ts`) dynamically from endpoint metadata — tests are generated output, not hand-written
- `SdkGenerator` instantiates `ReactQueryGenerator` and passes `inlinePathParameters` config along with other dependencies
- Updated `TypescriptProject`, `SimpleTypescriptProject`, and `BundledTypescriptProject` to accept `reactQueryExportPaths: string[]` (plural) and wire each into `package.json` exports
- Added `@tanstack/react-query` and `react` as optional peer dependencies when feature is enabled
- Added `@types/react` as devDependency when feature is enabled (needed for TypeScript compilation of generated `context.ts`)
- Added test dependencies (`@testing-library/react`, `react-dom`, `happy-dom`) as devDependencies when feature is enabled

### Generated Output Structure
```txt
src/react-query/
├── context.ts                        # ClientProvider component + useClient() hook
├── index.ts                          # barrel: export { ClientProvider, useClient }, export * as endpoints, ...
├── endpoints/
│   ├── index.ts                      # barrel: export * as container, export * as enum, ...
│   ├── container/index.ts            # getAndReturnListOfPrimitivesMutationOptions, useGetAndReturnListOfPrimitivesMutation, ...
│   ├── enum/index.ts
│   └── ...
├── noAuth/index.ts
├── noReqBody/index.ts
└── ...
tests/unit/react-query/              # Generated test files (produced by ReactQueryGenerator)
├── query-keys.test.ts
├── options-factories.test.ts
├── hooks.test.ts
├── context.test.ts
└── invalidation.test.ts
```

### Generated API surface (per endpoint)

**Naming convention**: All non-hook exports use **camelCase** (e.g. `listItemsOptions`, `listItemsQueryKey`). Hooks use `useXxx` / `useSuspenseXxx` / `useXxxMutation` PascalCase prefix per React convention.

**Client injection**: All generated hooks resolve the client from React context via `useClient()` — no `client` parameter needed. Options factories still accept an explicit `client` parameter for SSR/`prefetchQuery` compatibility.

```tsx
// App setup:
import { ClientProvider } from "@company/sdk/react-query";
<ClientProvider client={client}><App /></ClientProvider>

// In any component — no client param needed:
const { data } = useListItems(request);
const { mutate } = useCreateItemMutation();
```

For **GET/HEAD** endpoints:
- `xxxQueryKey(...)` — cache key factory; excludes `requestOptions` to prevent cache pollution
- `xxxOptions(client, ...)` — composable options factory wrapped in `queryOptions()` for idiomatic TanStack Query v5
- `useXxx(...)` — convenience hook with option overrides (client resolved from context)
- `useSuspenseXxx(...)` — suspense variant (client from context)
- `invalidateXxx(queryClient, ...)` — invalidate by request args (not requestOptions)
- `invalidateAllXxx(queryClient)` — only generated for endpoints **with parameters** (omitted for no-param endpoints where it would be identical to `invalidateXxx`)
- If paginated: `xxxInfiniteOptions(client, ...)` + `useXxxInfinite(...)` (client from context)

For **non-GET** endpoints:
- `xxxMutationOptions(client, requestOptions?)` — composable options factory with `mutationFn` signature aligned to `useMutation`'s `TVariables` pattern; accepts optional `requestOptions` for single-param and no-param endpoints
- `useXxxMutation(requestOptions?, options?)` — convenience hook (client from context) with option overrides and optional SDK `requestOptions`

> **Hook option overrides**: All generated hooks accept an optional `options` parameter that allows users to customize TanStack Query behavior (e.g., `enabled`, `staleTime`, `gcTime`, `select`, `retry`) without falling back to options factories. Options are typed with `Omit<>` to prevent accidental overrides of core config like `queryKey` and `queryFn`.

> **Query parameter handling** (consistent with mutations):
> - **Single-param** (sdkRequest, no path params): `request` separated from `requestOptions` for clean DX — users call `useXxx(request, requestOptions?, options?)`. QueryKey excludes `requestOptions` to prevent cache pollution.
> - **Multi-param** (has path params): Named parameters extracted from tuple — users call `useGetWithPath(param, requestOptions?, options?)` instead of opaque `useGetWithPath(args, options?)`. QueryKey excludes `requestOptions` to prevent cache pollution.
> - **No params**: Accepts explicit `requestOptions` — users call `useXxx(requestOptions?, options?)`.

> **Mutation parameter handling**:
> - **Single-param** (no path params): `TVariables = Params[0]` — users call `mutate(request)` directly. Hook/factory accept optional `requestOptions?: Params[1]` parameter for SDK options (timeout, abortSignal, headers).
> - **Multi-param** (has path params): `TVariables = Params` (full tuple) — users call `mutate([pathParam, body, requestOptions?])`. requestOptions is already part of the tuple.
> - **No params**: `TVariables = void` — users call `mutate()`. Hook/factory accept optional `requestOptions?: Params[0]` parameter for SDK options.
> - Both the mutation options factory and the hook use the **same `TVariables` pattern**, so the factory is safely composable with `useMutation()` directly.

### Seed Test
- Added `exhaustive:react-query` fixture — validates generated code compiles and passes tests
- Added `versions.yml` entry for v3.54.0
- **Generator-produced unit tests**: `ReactQueryGenerator` now includes `generateTestFiles()` method that dynamically generates 5 comprehensive unit test files (59 tests total) from endpoint metadata:
  - **Query keys**: Structure, uniqueness, requestOptions exclusion (13 tests)
  - **Options factories**: queryOptions/infiniteQueryOptions wrapping, consistency, requestOptions handling (22 tests)
  - **Hooks**: Client context resolution, parameter passing, option overrides (10 tests)
  - **Context/Provider**: useClient error handling, provider nesting (6 tests)
  - **Invalidation helpers**: QueryClient.invalidateQueries calls with correct cache key specificity (8 tests)
- Tests classify endpoints by type (no-param, single-param, multi-param, pagination) and generate representative test cases for each category
- Tests are regenerated automatically during seed fixture regeneration — no manual test file protection needed
- [ ] Updated README.md generator (not applicable)

## Updates Since Initial PR
- Added **hook wrappers** (`useQuery`, `useSuspenseQuery`, `useInfiniteQuery`, `useMutation`) alongside existing options factories
- Added **cache invalidation helpers** (`invalidateXxx`, `invalidateAllXxx`) for GET/HEAD endpoints
- Fixed **mutation hook** generation to support endpoints with **multiple positional params** by using the full `Parameters<...>` tuple for variables
- Fixed **endpoint reserved-word correctness** by using `endpoint.name.camelCase.unsafeName` for client method access/type extraction (while still using `safeName` for generated exported function identifiers)
- Avoid importing `QueryKey` in mutation-only service files
- **Implemented real `getNextPageParam`** using Fern's pagination metadata:
  - **CursorPagination**: Extracts `next` cursor from `lastPage.response` (e.g. `lastPage.response?.next`)
  - **OffsetPagination**: Increments offset using `lastPageParam + lastPage.data.length`, checks `hasNextPage` response property if available
  - **UriPagination/PathPagination**: Extracts `nextUri`/`nextPath` from response
  - **CustomPagination**: Falls back to stub returning `undefined` (cannot auto-implement)
  - `queryFn` now injects `pageParam` into the pagination request field (e.g. `cursor`, `offset`)
- **Added TanStack Query option overrides to all generated hooks**:
  - All hooks (`useQuery`, `useSuspenseQuery`, `useInfiniteQuery`, `useMutation`) now accept optional `options` parameter
  - Options use `Omit<>` types to prevent overriding core config (`queryKey`, `queryFn`, `initialPageParam`, `getNextPageParam`, `mutationFn`)
  - Hook signatures changed from `(client, ...args)` to `(client, args, options?)` for endpoints with request params
  - Options are spread over factory outputs: `{ ...factory(...), ...options }` so user overrides take precedence
- **Fixed mutation options factory `mutationFn` signature**: Factory's `mutationFn` now uses the same `TVariables` pattern as the hooks (single-param unwrapping, multi-param tuple, no-param void)
- **Added `requestOptions` parameter support for single-param and no-param mutations/queries**
- **Skipped `invalidateAll` generation for no-param GET endpoints**
- **Added React Context/Provider pattern for idiomatic client injection**
- **Renamed options factories and query key functions to camelCase**
- **Fixed multi-param paginated infinite query destructuring bug**
- **Use `queryOptions()` / `infiniteQueryOptions()` from TanStack Query v5**
- **Multi-param query hooks now use named parameters instead of opaque tuples**
- **Fixed path parameter inlining detection logic** (critical bug fix)
- **Implemented generator-based test generation** (latest update):
  - Added `generateTestFiles()` method to `ReactQueryGenerator` that produces test files dynamically from endpoint metadata
  - Tests are now generated output (written via `rootDirectory.createSourceFile()`) rather than hand-written protected files
  - Removed `.fernignore` — no longer needed since tests naturally survive seed regeneration by being generator output
  - Test generation uses endpoint classification heuristics (no-param/single-param/multi-param, pagination status) to generate representative test cases
  - Fixed biome lint errors (line length, non-null assertions) in test generation code

## ⚠️ Items for Reviewer Attention
1. **String-based code generation** — `ReactQueryGenerator` builds output files (including tests) via `lines.push(...)` string concatenation rather than using the AST/ts-morph approach. This is functional but fragile. The test generation adds ~1200 additional lines to an already large generator class.
2. **Generated test comprehensiveness** — Tests are now generated from endpoint metadata using classification heuristics. While they cover all major patterns, they may miss edge cases that hand-written tests would catch. The trade-off is that generated tests automatically stay in sync with the generated React Query API surface.
3. **Import path assumptions** — generated files import `Client` via relative paths (e.g. `"../../../Client"`). This assumes `relativePackagePath` is `"src"`. If customized, imports may break.
4. **`pageParam as never` cast** — To support injecting `pageParam` (which is `unknown`) into typed request objects, the generator casts `pageParam as never`. This silences TypeScript but could mask runtime type mismatches if pagination field types don't match.
5. **Offset pagination `lastPageParam` dependency** — Offset pagination uses `lastPageParam` from TanStack Query's callback to track the current offset. This assumes TanStack Query properly threads this parameter through (which it does as of v5).
6. **URI/Path pagination assumes `"cursor"` field name** — These pagination types hardcode `requestFieldName: "cursor"` rather than reading from the IR's `page.property.name`. This may not match all API definitions.
7. **Identical ternary branches in `requestFieldName` extraction** — The code checks `pagination.page.property.type === "query"` but both branches return the same value. This suggests the code may not differentiate between query params and body properties for pagination fields.
8. **Query key cache key structure** — All queries exclude `requestOptions` from cache key (good for preventing pollution). This is consistent across single-param and multi-param queries.
9. **Services with both endpoints and children (edge case)** — The `_service.ts` pattern handles services that have both their own endpoints and child sub-services; this is defensive but not covered by the `exhaustive` fixture.
10. **Root-level service naming** — Root services are placed in `react-query/root/`, which could theoretically conflict if a real service is named `"root"`.
11. **requestOptions binding time** — For single-param, no-param mutations/queries, `requestOptions` are bound at hook/factory creation time (not per-call). This is appropriate for options that are typically consistent per hook instance (timeout, abortSignal, headers), but means they cannot vary between calls to `mutate()` or query refetches.
12. **Params index extraction for requestOptions** — For single-param queries, `requestOptions` is typed as `Params[1]`; for no-param queries, it's `Params[0]`; for multi-param, it's `Params[pathParamNames.length + (hasRequestBody ? 1 : 0)]`. Verify these correctly index into the SDK's RequestOptions type in all cases.
13. **`@types/react` version range** — Currently set to `>=18.0.0` which is very broad. Future React type definition breaking changes could cause compilation issues. Consider tightening to `^18.0.0` or `>=18.0.0 <20.0.0`.
14. **Two different APIs for client param** — Hooks use context (no `client` param), while options factories require explicit `client` param. This is intentional for SSR compatibility but creates two patterns users need to learn.
15. **`useClient()` error message** — Throws with clear error if used outside `<ClientProvider>`. Error message is helpful but could be confusing if users don't realize hooks require the Provider.
16. **Multi-param paginated infinite query fix not directly tested** — The `exhaustive` seed fixture doesn't contain any paginated GET endpoints with separate path parameters, so the multi-param pagination fix codepath is not exercised by the seed tests.
17. **`queryOptions()` / `infiniteQueryOptions()` workaround for `isolatedDeclarations`** — The return type of these helpers contains internal `DataTag` symbols that cannot be named. The workaround is to call the helpers at runtime (for idiomacy) but provide explicit simplified return type annotations. This relies on the helpers being identity functions at runtime.
18. **Multi-param named parameters use index-based types** — Multi-param hooks/factories use types like `param: Params[0]`, `requestOptions?: Params[1]`. If the SDK method signature changes parameter ordering, this could silently break.
19. **🔴 CRITICAL: Path param detection logic duplicated from `RequestWrapperContextImpl`** — The React Query generator replicates the `shouldInlinePathParameters()` logic instead of sharing it. If the original logic changes, the generator could get out of sync.
20. **🟡 Test dependencies added to ALL generated React Query SDKs** — The generator adds `@testing-library/react`, `react-dom`, and `happy-dom` as devDependencies whenever `generateReactQuery: true`, meaning ALL customer SDKs with React Query enabled will have these test libraries even though they won't have test files (only seed fixtures have tests). This is unnecessary bloat for customer SDKs.
21. **🟡 NEW: Generated test fragility** — The test generation code uses string concatenation to build test files. Any changes to generated React Query API surface (function names, signatures, import paths) could cause generated tests to fail compilation or produce incorrect assertions. Consider whether a more robust approach (e.g., ts-morph AST manipulation) would be safer for test generation.

## Human Review Checklist
- [ ] Verify the mutation options factory `mutationFn` signature matches the hooks correctly (single-param unwrapping, multi-param tuple, no-param void)
- [ ] Verify `requestOptions` parameter support for single-param and no-param mutations/queries works as intended
- [ ] Confirm that `Params` type extraction for mutations/queries correctly identifies requestOptions type for all cases
- [ ] Verify QueryKey for all queries correctly excludes `requestOptions` from cache key (prevents cache pollution)
- [ ] Review whether binding `requestOptions` at hook creation time (rather than per-call) is acceptable DX
- [ ] Verify the hook signature patterns provide good DX
- [ ] Review the `Omit<>` types to ensure they correctly match TanStack Query v5's type signatures
- [ ] Verify the factory is safely composable with `useMutation()` and `useQuery()` directly for all parameter patterns
- [ ] Verify `invalidateAll` is correctly omitted for no-param GET endpoints and present for parameterized endpoints
- [ ] Consider whether string-based code generation (including test generation) should be migrated to ts-morph pattern
- [ ] Verify Context/Provider pattern works as expected
- [ ] Verify `useClient()` error message is clear when Provider is missing
- [ ] Review camelCase naming convention
- [ ] Consider whether `@types/react` version range should be tightened (currently `>=18.0.0`)
- [ ] **Verify the multi-param pagination fix**: Confirm that `endpoint.allPathParameters.length` correctly identifies the request object index for all SDK method signatures
- [ ] **Verify `queryOptions()` / `infiniteQueryOptions()` usage**: Confirm that wrapping options in these helpers while using explicit return type annotations is safe and idiomatic
- [ ] **Verify multi-param named parameters**: Confirm that index-based types correctly map to path params and requestOptions
- [ ] **🔴 CRITICAL: Verify path param detection logic matches `RequestWrapperContextImpl.shouldInlinePathParameters()`**: Confirm the three-condition check is correct and consider refactoring to share logic
- [ ] **🟡 Review test dependency strategy**: Decide whether test deps should be added to ALL generated React Query SDKs or only to seed fixtures
- [ ] **🟡 NEW: Review generated test quality**: Examine the generated test files in the seed fixture to verify they provide adequate coverage of the generated React Query API surface. Consider whether endpoint classification heuristics miss any important test cases.
- [ ] **🟡 NEW: Verify test generation robustness**: Confirm that generated tests will remain valid if the React Query generator code changes (e.g., function name refactors, signature changes). Consider whether a more structured approach would be safer than string concatenation.

## Testing
- [x] `pnpm run check` (biome)
- [x] Seed fixture: `exhaustive:react-query` passes generation + build + tests
- [x] TypeScript compilation verified locally (`npx tsc --noEmit` passes with zero errors)
- [x] All generated patterns verified in seed output
- [x] **NEW**: All 59 generated unit tests passing locally (389 total tests including wire tests)
- [x] **NEW**: Test generation produces valid TypeScript that compiles without errors
- [x] **NEW**: Generated tests correctly import from generated React Query code
- [x] **NEW**: Biome lint passes on generated test code (formatting, no non-null assertions)
- [ ] Multi-param paginated infinite query with path params (not covered by `exhaustive` fixture — no test case exists)
- [ ] Manual testing completed (not applicable; generator-only)

## CI Status
All required checks passing (6/6 pass: compile, lint, depcheck, biome, test, Validate versions.yml). The `test-ete` failure is a known flaky npm/migration-cache error unrelated to these changes.

---

<a href="https://app.devin.ai/review/fern-api/fern/pull/13057" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: generators/typescript/sdk/generator/src/SdkGenerator.ts [hunk 1]
```diff
@@ -2031,6 +2053,187 @@ export class SdkGenerator {
...
+        if (endpointEntries.length === 0) {
+            return undefined;
+        }
+
+        // Build the react-query/index.ts file content
+        const lines: string[] = [];
+        lines.push(`// This file was auto-generated by Fern from our API Definition.`);
+        lines.push(``);
+        lines.push(`import type { QueryKey } from "@tanstack/react-query";`);
+        lines.push(`import { ${rootClientName} } from "../Client";`);
```
[the remark this answers is not in the document]
Comment (fern-support) [reply generators/typescript/sdk/generator/src/SdkGenerator.ts]: yes, apply

File: generators/typescript/sdk/generator/src/SdkGenerator.ts [hunk 2]
```diff
@@ -2031,6 +2053,187 @@ export class SdkGenerator {
...
+        lines.push(``);
+
+        // Type alias for the client
+        lines.push(`type ClientInstance = InstanceType<typeof ${rootClientName}>;`);
+        lines.push(``);
+
+        for (const { endpoint, servicePath } of endpointEntries) {
+            const endpointName = endpoint.name.camelCase.safeName;
+            const isGetMethod = endpoint.method === "GET";
```
[the remark this answers is not in the document]
Comment (fern-support) [reply generators/typescript/sdk/generator/src/SdkGenerator.ts]: Agreed

**Review by fern-support — CHANGES_REQUESTED** [review]
